### PR TITLE
Skip subnets that contain no tags to bypass KeyError

### LINKED ISSUE #44330: salt-cloud ec2 driver subnetname to subnetid translation bug
### Description of Issue/Question

Running salt-cloud with a subnetname in the profile will fail if not all subnets have tags.
In salt/cloud/clouds/ec2.py the function def: def _get_subnetname_id(subnetname), a KeyError will be raised on this check:

if tag['key'] == 'Name' and tag['value'] == subnetname:

This is because tag might not be a list or dict, and might not have anything in there at all. This happens because aws.query a few lines up returns a collection of subnets, but not every subnet has tags, i.e. default subnets in a default VPC have no tags at all. This causes tagSet to be non-existent, which in turn causes the tags variable to be nothing.

Adding a quick check to make sure tagSet actually exists before continuing fixes this for me. The def then looks like this:

````
    '''
    Returns the SubnetId of a SubnetName to use
    '''
    params = {'Action': 'DescribeSubnets'}
    for subnet in aws.query(params, location=get_location(),
               provider=get_provider(), opts=__opts__, sigver='4'):
        tags = subnet.get('tagSet', {}).get('item', {})
        if(subnet.has_key("tagSet")):
            if not isinstance(tags, list):
                tags = [tags]
            for tag in tags:
                if tag['key'] == 'Name' and tag['value'] == subnetname:
                    log.debug('AWS Subnet ID of {0} is {1}'.format(
                        subnetname,
                        subnet['subnetId'])
                    )
                    return subnet['subnetId']
    return None
````
The added line was if(subnet.has_key("tagSet")): and I indented everything below it to skip that specific loop run.

I'm not 100% sure what this is supposed to help with:
````
            if not isinstance(tags, list):
                tags = [tags]
````
perhaps it was an earlier attempt at detecting the existence, but instead of breaking the loop it created a list of whatever tags was.

### Versions Report
````
Salt Version:
           Salt: 2017.7.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.10 (default, Jul 15 2017, 17:16:57)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.2
            ZMQ: 4.2.2

System Versions:
           dist:
         locale: UTF-8
        machine: x86_64
        release: 17.2.0
         system: Darwin
        version: 10.13.1 x86_64
````

### PR DESCRIPTION
**Pull request by bwaang**
### What does this PR do?
When using `subnetname` in salt-cloud profile, while looping if hits a subnets with no tags cause this error:

```bash
[ERROR   ] Failed to create VM new-ec2-instance. Configuration value 'key' needs to be set
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/salt/cloud/__init__.py", line 1253, in create
    output = self.clouds[func](vm_)
  File "/usr/lib/python3.7/site-packages/salt/cloud/clouds/ec2.py", line 2645, in create
    data, vm_ = request_instance(vm_, location)
  File "/usr/lib/python3.7/site-packages/salt/cloud/clouds/ec2.py", line 1915, in request_instance
    _new_eni = _create_eni_if_necessary(interface, vm_)
  File "/usr/lib/python3.7/site-packages/salt/cloud/clouds/ec2.py", line 1434, in _create_eni_if_necessary
    interface['SubnetId'] = _get_subnetname_id(interface['subnetname'])
  File "/usr/lib/python3.7/site-packages/salt/cloud/clouds/ec2.py", line 1247, in _get_subnetname_id
    if tag['key'] == 'Name' and tag['value'] == subnetname:
KeyError: 'key'
```
Since it's assumed we want to match a `subnetname` we can safely bypass tags with no tags since it's completely null.  Many AWS default subnets do not have tags which cause this error to occur if the empty subnet tags appear earlier in the returned list from the `DescribeSubnets` AWS REST call.

This PR checks first for the existence of the `tagSet` before traversing the `tagSet` and will bypass if the `tagSet` does not exist for a returned subnet

### What issues does this PR fix or reference?
Fixes: #44330 - This still occurs in the latest Saltstack `3003`

### Previous Behavior
Using `subnetname` in salt-cloud profile will generate a `KeyError` whenever it hits a subnet with no tags.

### New Behavior
Using `subnetname` will not throw a `KeyError` and can be readily used without error if any subnets with not tags exists.

### Merge requirements satisfied?
- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by Ch3LL — DISMISSED** [review]

**Comment by Ch3LL** [comment]
pre-commit needs to be fixed and we should be able to get this merged.

**Comment by bwaang** [comment]
Looks like the `pre-commit` passed this time :)

**Review by s0undt3ch — DISMISSED** [review]

**Review by Ch3LL — DISMISSED** [review]

**Review by garethgreenaway — DISMISSED** [review]

File: salt/cloud/clouds/ec2.py [hunk 1]
```diff
@@ -1223,13 +1223,16 @@ def _get_subnetname_id(subnetname):
...
     ):
-        tags = subnet.get("tagSet", {}).get("item", {})
-        if not isinstance(tags, list):
-            tags = [tags]
-        for tag in tags:
-            if tag["key"] == "Name" and tag["value"] == subnetname:
-                log.debug("AWS Subnet ID of %s is %s", subnetname, subnet["subnetId"])
-                return subnet["subnetId"]
+        if "tagSet" in subnet:
+            tags = subnet.get("tagSet", {}).get("item", {})
```
Comment (waynew) [inline salt/cloud/clouds/ec2.py:1227]: This `if` is already effectively happening with `subnet.get("tagSet")`. The actual problem is happening on line 1231 - apparently there is a key *in* `subnet["tagSet'"]["item"]` that's missing a `"key"`. I don't know much more about what's in the ec2 data than that, but unless I'm missing something, I do not see how this fix will change anything?

**Review by waynew — CHANGES_REQUESTED** [review]

File: salt/cloud/clouds/ec2.py [hunk 2]
```diff
@@ -1223,13 +1223,16 @@ def _get_subnetname_id(subnetname):
...
+            if not isinstance(tags, list):
+                tags = [tags]
+            for tag in tags:
+                if tag["key"] == "Name" and tag["value"] == subnetname:
```
Comment (waynew) [inline salt/cloud/clouds/ec2.py:1231]: This is the line that's raising the exception. I'm not sure what is *in* tag... but clearly it was missing a `"key"`.

A fix that *seems* correct (without looking into it deeper, I'm not 100% confident) would be to change this:

```
if tag.get('key') == "Name" and tag.get("value") == subnetname:
```

This does assume that the subnetname is going to be non-None, which may be incorrect. If `subnetname` could ever be none, `tag.get("value", object()) == subnetname` would be better, since subnetname will always be something not a brand new object.

**Review by twangboy — DISMISSED** [review]
The windows test failures are unrelated.

File: salt/cloud/clouds/ec2.py (hunk 2 above)
Comment (bwaang) [reply salt/cloud/clouds/ec2.py:1231]: @waynew I've debugged this using `--log-level=trace` and if there's no tags whatsoever, you end up with the complete absence of `tagSet` completely.   The AWS data model seems to just not return tagSet at all if you have no tags whatsoever.  So the solution proposed would end up trying to do a `.get` on an empty dictionary `{}` yielding a similar error.

Alternatively we can look for the `tagSet` key, but because of `pep8` standard it doesn't like the `.has_key()` so ended up having to use `in`

Hope that makes sense, I've tested this fix and it works in skipping the subnets that are missing the `tagSet` key completely (which is an okay assumption, if you have no `tagSet`, you then don't have a `Name` tag).  Generally the AWS defaults don't come with `Name` tags.

Comment (waynew) [reply salt/cloud/clouds/ec2.py:1231]: Oh, haha. This is exactly why we require writing tests. I just threw together a quick example to try and illustrate how they're the same...

```
def as_written(subnet, subnetname='bob'):
    tags = subnet.get("tagSet", {}).get("item", {})
    if not isinstance(tags, list):
        tags = [tags]
    for tag in tags:
        assert False, "We had tags, what?"
        if tag["key"] == "Name" and tag["value"] == subnetname:
            return subnet["subnetId"]


def as_changed(subnet, subnetname='bob'):
    if "tagSet" in subnet:
        tags = subnet.get("tagSet", {}).get("item", {})
        if not isinstance(tags, list):
            tags = [tags]
        for tag in tags:
            assert False, "We had tags, what?"
            if tag["key"] == "Name" and tag["value"] == subnetname:
                return subnet["subnetId"]


data = {"no tagSet here": "sersiously, there isn't one"}
as_written(data)
as_changed(data)

other_data = {"tagSet": "there is a tagset, but no key in any of these things"}

try:
    as_written(other_data)
except AssertionError:
    print('oops as written does not work')

try:
    as_changed(other_data)
except AssertionError:
    print('oops as changed does not work')
```

Except... they're not. Because a little check:

```
def as_written(subnet, subnetname='bob'):
    tags = subnet.get("tagSet", {}).get("item", {})
    if not isinstance(tags, list):
        print('lol, empty dict is not a list')
        tags = [tags]
        print('here are tags:', tags)
```

What's happening is that when tagSet does not exist, neither does item, so the *default* value is `{}`. And `{}` isn't an instance of list, doh!

It looks like the correct fix is just 

```
    tags = subnet.get("tagSet", {}).get("item", {})
    # to this ----------------------------------^v
    tags = subnet.get("tagSet", {}).get("item", [])
```

Subtle change, but that's the kind of thing that gets exposed when tests are written to pass this function:

1. an empty dictionary
2. a dictionary with "tagSet", but is empty
3. a dictionary with tagSet but the item is not a list (I guess it would be just a single dict with key and value
4. a dictionary with tagSet and the item is a list of items with key/value

And of course, the subnet is gonna need "subnetId" when the key/value matches.

If you need more help - attend one of the [Test Clinics on Twitch](https://www.twitch.tv/saltstackinc/schedule) and I'll be glad to clarify any questions! (Also, they are recorded so if you're unable to make that time, let me know and I can also try to record something a little more async.

HTH!

Comment (bwaang) [reply salt/cloud/clouds/ec2.py:1231]: @waynew thanks I went ahead and tested, both solutions work (what I had written and yours).  If this is going to block the fix I'll just swap to yours.

Maybe we should just get rid of that check altogether, if we go with what I added to reduce confusion?  I just didn't want to touch it unnecessarily.  So it looks like it's creating an empty tagSet `{}` with a tag empty tag array `[]` that will just get bypassed in the next loop.

```python
        tags = subnet.get('tagSet', {}).get('item', {})
        if not isinstance(tags, list):
            tags = [tags]
```

But what I have works as well, it just completely bypasses the case where there is no `tagSet`, here's an example output from the AWS `DescribeSubnets` (with sensitive data masked).  The first one has no name, so `tagSet` completely doesn't exist so doing `if "tagSet" in subnet:` will just bypass that case.

```xml
        <item>
            <subnetId>subnet-*******</subnetId>
            <state>available</state>
            <vpcId>vpc-********</vpcId>
            <cidrBlock>********</cidrBlock>
            <availableIpAddressCount>250</availableIpAddressCount>
            <availabilityZone>********</availabilityZone>
            <defaultForAz>false</defaultForAz>
            <mapPublicIpOnLaunch>false</mapPublicIpOnLaunch>
        </item>
        <item>
            <subnetId>subnet-********</subnetId>
            <state>available</state>
            <vpcId>vpc-********</vpcId>
            <cidrBlock>********</cidrBlock>
            <availableIpAddressCount>245</availableIpAddressCount>
            <availabilityZone>********</availabilityZone>
            <defaultForAz>false</defaultForAz>
            <mapPublicIpOnLaunch>false</mapPublicIpOnLaunch>
            <tagSet>
                <item>
                    <key>Name</key>
                    <value>Subnet with a Name</value>
                </item>
            </tagSet>
        </item>
```

Salt-cloud has been great, just really want get this fixed so it no longer blocks us from using `subnetname` rather than having to hardcode and create mappings to `SubnetId` in our IAC.   Will make our automation a lot smoother!

Comment (bwaang) [reply salt/cloud/clouds/ec2.py:1231]: Actually went ahead and incorporated your suggestion as well @waynew since there's probably other corner cases that this covers & that current line is broken anyways.

**Review by Ch3LL — CHANGES_REQUESTED** [review]
Can we add a test. The work @waynew did does show the need for a test to cover these corner cases. And so we do not regress in the future.

**Comment by bwaang** [comment]
@Ch3LL @waynew @sagetherage added shiny new test ✨ tested it out, if you revert the code you'll get the same nasty stack trace as before keeping the test if you run:

```bash
nox -e "pytest-zeromq-3(coverage=False)" -- tests/unit/cloud/clouds/test_ec2.py
```

But with the new code everything passes 😁 

Hopefully now ready to merge 🤞

**Review by Ch3LL — APPROVED** [review]

**Comment by bwaang** [comment]
Hi @waynew just checking if need your final re-review on this, wasn't sure what are next steps now to get this merged and on the track to release.

I also noticed this fix was attempted several times, just wanted to make sure I wasn't missing anything to get it fixed:
https://github.com/saltstack/salt/pull/53113/files
https://github.com/saltstack/salt/pull/44363/files
https://github.com/saltstack/salt/pull/44362/files

**Comment by Ch3LL** [comment]
yep, we just need your re-review @waynew and all passing tests to get this merged in. I updated the branch because there was a fix put in to fix the macosx failures.

**Review by waynew — APPROVED** [review]

**Comment by bwaang** [comment]
Thanks @Ch3LL & @waynew!

**Comment by bwaang** [comment]
Hmm @Ch3LL looks like the mac tests passed now, but the CentOS7 ones only failed because it failed to spin up the workspace in the Jenkins?  Thoughts?

**Comment by Ch3LL** [comment]
I restarted the failing ones. I'll monitor it and hopefully get this merged in by tomorrow. thanks

**Comment by bwaang** [comment]
@Ch3LL ah :( feel like we're playing russian roulette a different one timed out this time, now it's the `freebsd` one.

Thanks for managing all these PRs btw, know it's probably a lot to manage

**Comment by Ch3LL** [comment]
I apologize for the flaky nature of the tests. We have made a lot of progress on the flaky behavior but still have a bit more to go. Thanks the reminder on this one :)

**Comment by bwaang** [comment]
No worries :), it's good these test are thorough, I'm quite impressed how many different OS this will make sure to cover regression.

Just FYI looks like `centos-7` failed this time, but passed last time :(

@Ch3LL I know this might be a funny ask & actually I don't know who has the power to do this, but do you think I would be able to trigger the ability to re-run the tests?  This way, I can then monitor it for you then when they all pass I can let you merge it so then you don't have to keep checking it every few days.  I'm sure you have hundreds to PRs to review & manage, can at least help you a bit on this one ;)

**Comment by Ch3LL** [comment]
Unfortunately I believe its blocked by permissions. I just restarted them. I apologize. We are working on these failing tests and they are definitely much better than they were a month ago.

**Comment by bwaang** [comment]
@Ch3LL no worries, that's what I figured - thought it wouldn't hurt to just ask :)

**Comment by bwaang** [comment]
@Ch3LL all tests passed! 🎉  🥳  🎊

**Comment by Ch3LL** [comment]
yay! Thanks for your patience :)
